### PR TITLE
add: semantic segmentation using PT and Trainer.

### PR DESCRIPTION
**Pull request by sayakpaul**
Just an approximate port of the TF variant: https://github.com/huggingface/notebooks/blob/main/examples/semantic_segmentation-tf.ipynb

I have used some text from that notebook as well as some from the blog post: https://huggingface.co/blog/fine-tune-segformer. The code is all from the blog post. 

The idea behind adding notebook is to attain some consistency in how the `examples` are organized and structured in this repository. Plus the blog post has additional points on the data labeling process (which is great) and the notebook introduced in this PR discards those to keep the content lightweight and more focused. 

Closes https://github.com/huggingface/notebooks/issues/267

### CODE REVIEW

File: examples/semantic_segmentation.ipynb [hunk 1]
```diff
@@ -0,0 +1,1253 @@
+{
```
Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>"This notebook shows" -&gt; "In this notebook, you will learn how to fine-tune a pretr..."</p><p> </p><p>Otherwise wording is same as first sentence, so sounds like you're repeating yourself.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>First sentence here feels like the 1st sentence of the notebook and can probably be removed. Maybe move 2nd sentence here to first markdown cell too and just delete this one.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>Maybe add this? (nit, up to you to change if you agree)</p><p> </p><p>"... version of those libraries or run the <code>pip install</code> command above with the <code>--upgrade</code> flag"</p><p>  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>First sentence here again repeating yourself 😅</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>Move this to the markdown cell above this so the info comes before you call <code>load_dataset</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>Maybe something worded like this?</p><p> </p><p>"Specifically, we'll want to mask the pixels where the network predicted <code>unlabeled</code> and avoid computing the loss for it since it doesn't contribute to to training that much" </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>Maybe add a comment along the lines of:</p><p> </p><p>"We'll explicitly define a random seed to use when calling <code>ds.shuffle</code> to ensure our results are the same each time we run this cell."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>Maybe an explanation above this cell explaining the difference between train and val transforms and (optionally, if you think necessary) how the feature extractor is being used in both?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>"This will make sure a custom segmentation head is created" (will be -&gt; is)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>"We need to ..." -&gt; "To use the <code>Trainer</code>, we'll need to define the training configuration and any evaluation metrics we might want to use".   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

Comment (sayakpaul) [reply examples/semantic_segmentation.ipynb:1]: <p>Kind of conflicted here :) </p><p> </p><p>I am trying to not break the consistency of the other notebooks (such as https://github.com/huggingface/notebooks/blob/main/examples/image_classification.ipynb) in terms of how content is outlined. </p><p> </p><p>But I do see your point. </p>

Comment (sayakpaul) [reply examples/semantic_segmentation.ipynb:1]: <p>I think the second part is already covered? </p><p> </p><blockquote>To make sure we (1) resize to the appropriate size (2) use the appropriate image mean and standard deviation for the model architecture we are going to use, we instantiate what is called a feature extractor with the <code>AutoFeatureExtractor.from_pretrained</code> method. </blockquote>

[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/semantic_segmentation.ipynb:1]: <p>Absolutely right. </p>

[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/semantic_segmentation.ipynb:1]: <p>Point noted. Will incorporate it in the next tutorial, for sure. </p>

Comment (nateraw) [reply examples/semantic_segmentation.ipynb:1]: Ah I see...its just my opinion. feel free to stick with it if it's following a pattern you like

Comment (sayakpaul) [reply examples/semantic_segmentation.ipynb:1]: And I second you. The latest updates I am about to push will get rid of it :)

**Comment by sayakpaul** [comment]
@nateraw thanks for your comprehensive comments. I had to reflect that in my commit message. 

Let me know your thoughts on the latest updates.

File: examples/semantic_segmentation.ipynb [hunk 2]
```diff
@@ -0,0 +1,1266 @@
...
```
Comment (nateraw) [inline examples/semantic_segmentation.ipynb:1]: <p>Change "Here's a call for actions" to something like...</p><p> </p><p>If you want to explore further beyond this notebook, here are some things you can try next:</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/287/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='57'/>

**Review by nateraw — APPROVED** [review]
Other than above comment about call to action, this looks good to go to me

File: examples/semantic_segmentation.ipynb (hunk 2 above)
Comment (sayakpaul) [reply examples/semantic_segmentation.ipynb:1]: Sounds way better. Thanks.
